tools/lkl/lib/virtio_net.c

```diff
@@ -76,15 +114,47 @@ void poll_thread(void *arg)
 	struct virtio_net_poll *np = (struct virtio_net_poll *)arg;
 	int ret;
 
+	/* Synchronization is handled in virtio_process_queue */
 	while ((ret = np->dev->ops->poll(np->dev->nd, np->event)) >= 0) {
 		if (ret & LKL_DEV_NET_POLL_RX)
 			virtio_process_queue(&np->dev->dev, 0);
 		if (ret & LKL_DEV_NET_POLL_TX)
 			virtio_process_queue(&np->dev->dev, 1);
-		lkl_host_ops.sem_down(np->sem);
 	}
 }
 
+static void free_queue_locks(struct lkl_mutex_t **queues, int num_queues)
+{
+	int i = 0;
+	if (!queues)
+		return;
+
+	for (i = 0; i < num_queues; i++)
+		lkl_host_ops.mem_free(queues[i]);
+
+	lkl_host_ops.mem_free(queues);
+}
+
+static struct lkl_mutex_t **init_queue_locks(int num_queues)
+{
+	int i;
+	struct lkl_mutex_t **ret = lkl_host_ops.mem_alloc(
+		sizeof(struct lkl_mutex_t*) * num_queues);
+	if (!ret)
+		return NULL;
+
+	memset(ret, 0, sizeof(struct lkl_mutex_t*) * num_queues);
+	for (i = 0; i < num_queues; i++) {
+		ret[i] = lkl_host_ops.mutex_alloc();
+		if (!ret[i]) {
+			free_queue_locks(ret, num_queues);
```

Comment:
I guess you can pass `i` instead of `num_queues` so you don't need to do the `memset`?